## Homework
The goal of this homework is to familiarize users with monitoring for ML batch services, using PostgreSQL database to store metrics and Grafana to visualize them.

## Q1. Prepare the dataset
Start with baseline_model_nyc_taxi_data.ipynb. Download the March 2024 Green Taxi data. We will use this data to simulate a production usage of a taxi trip duration prediction service.

#### load a data

In [8]:
import requests 
import datetime
import pandas as pd

from joblib import load, dump
from tqdm import tqdm

In [13]:
files = [('green_tripdata_2024-03.parquet', '../data'), ('green_tripdata_2024-02.parquet', '../data')]

print('Download files:')

for file, path in files:
    url=f'https://d37ci6vzurychx.cloudfront.net/trip-data/{file}'
    resp=requests.get(url, stream=True)
    save_path=f'{path}/{file}'
    with open(save_path, "wb") as handle:
        for data in tqdm(resp.iter_content(),
                        desc=f'{file}',
                        postfix=f'save to {save_path}',
                        total=int(resp.headers['Content-Length'])):
            handle.write(data)


Download files:


green_tripdata_2024-03.parquet: 100%|█| 1372372/1372372 [00:09<00:00, 138864.52it/s, save to ../data
green_tripdata_2024-02.parquet: 100%|█| 1283805/1283805 [00:09<00:00, 142423.75it/s, save to ../data


In [72]:
df_03 = pd.read_parquet('../data/green_tripdata_2024-03.parquet')

In [73]:
df_03.shape

(57457, 20)

What is the shape of the downloaded data? How many rows are there?

- 72044
- 78537
- 57457
- 54396

### Answer:
- 57457

## Q2. Metric
Let's expand the number of data quality metrics we’d like to monitor! Please add one metric of your choice and a quantile value for the "fare_amount" column (quantile=0.5).

Hint: explore evidently metric ColumnQuantileMetric (from evidently.metrics import ColumnQuantileMetric)

What metric did you choose?

- ColumnMissingValuesMetric

## Q3. Monitoring
Let’s start monitoring. Run expanded monitoring for a new batch of data (March 2024).

What is the maximum value of metric quantile = 0.5 on the "fare_amount" column during March 2024 (calculated daily)?

- 10
- 12.5
- 14.2
14.8